README.md
# mondo-clj

<img src="https://cdn.rawgit.com/adamneilson/mondo-clj/master/resources/mondo-logo.svg" style="width:60%;display:block;margin:auto;">

This **unofficial** clojure lib gives simple hooks for the [Mondo API](https://getmondo.co.uk/docs/). It's pre-pre-alpha as I'm a couple of days away from getting my grubby paws on my shiny new Mondo account. So as of yet it's not been tested in the slightest.

## Installing 
Current [semantic](http://semver.org/) version:

[![Clojars Project](http://clojars.org/mondo-clj/latest-version.svg)](http://clojars.org/mondo-clj)

To use with Leiningen, add

```clojure
:dependencies [[mondo-clj 0.0.5]]
```
to your project.clj.

You can use it in a source file like this:

```clojure
(:require [mondo-clj.core :as mondo])
```
or by REPL:

```clojure
(require '[mondo-clj.core :as mondo])
```


## Usage

### Authentication
The Mondo API implements OAuth 2.0 so the first step is to get a token:

```clojure
(mondo/get-access-token client-id
                        client-secret
                        username
                        password)
```

This should return a map similar to this:

```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :access-token "access_token"
 :client-id "client_id"
 :expires-in 21600
 :refresh-token "refresh_token"
 :token-type "Bearer"
 :user-id "user_id"}
```
As you can see I'm adding the `:status` to the returned map to give a bit of context to any errors.

At any time you can get information about an access token, using:

```clojure
(mondo/whoami access-token)
```

returns:

```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :authenticated true,
 :client-id "client_id",
 :user-id "user_id"}
```

### Accounts

To get a list of accounts:

```clojure
(mondo/list-accounts access-token)
```
Returns something like:

```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :accounts [{:id "acc_00009237aqC8c5umZmrRdh"
             :description "Peter Pan's Account"
             :created #inst "2015-12-01T00:00:00.000-00:00"}]}

```

### Balance
Get the balance of a specific account

```clojure
(mondo/read-balance access-token account-id)
```
Returns:

```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :balance 50.0
 :currency "GBP"
 :spend-today 0.0}
```
NB: The [Mondo docs](https://getmondo.co.uk/docs/#balance) currently say that the `balance` and `spend-today` return types are 64bit integers in minor units of the currency. Eg. GBP `5000` is £50.00. This seems counter-intuitive to me so I'll be coercing the `balance` and `spend-today` to *double* type so `50.0` is £50.00.

### Transactions
Get details of a single transaction. NB: full merchant details are returned:

```clojure
(mondo/get-transaction access-token transaction-id)
```

Returns 
```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :transaction {:account-balance 130.13
               :amount -5.10
               :created #inst "2015-12-01T00:00:00.000-00:00"
               :currency "GBP"
               :description "THE DE BEAUVOIR DELI C LONDON GBR"
               :id "tx_00008zIcpb1TB4yeIFXMzx"
               :merchant {:address {:address "98 Southgate Road"
                                    :city "London"
                                    :country "GB",
                                    :latitude 51.54151
                                    :longitude -0.08482400000002599
                                    :postcode "N1 3JD"
                                    :region "Greater London"}
                          :created #inst "2015-12-01T00:00:00.000-00:00"
                          :group_id "grp_00008zIcpbBOaAr7TTP3sv"
                          :id "merch_00008zIcpbAKe8shBxXUtl"
                          :logo "https://pbs.twimg.com/profile_images/527043602623389696/68_SgUWJ.jpeg"
                          :emoji "🍞"
                          :name "The De Beauvoir Deli Co."
                          :category "eating_out"}
               :metadata {}
               :notes "Salmon sandwich 🍞"
               :is_load false
               :settled true}}
```

Or list transactions

```clojure
(mondo/list-transactions access-token account-id)

;or 

(mondo/list-transactions access-token account-id {:since #inst "2015-12-01T00:00:00.000-00:00"})

;or 

(mondo/list-transactions access-token account-id {:before #inst "2015-12-01T00:00:00.000-00:00"})

;or 

(mondo/list-transactions access-token account-id {:since #inst "2015-12-01T00:00:00.000-00:00" :limit 48})
```

You can also add meta-data to a transaction

```clojure
(mondo/annotate-transaction access-token transaction-id {:foo "bar" :baz "quux"})
```


### Feed

You can inject items into an accounts feed

```clojure
(mondo/create-feed-item {:access-token "0000000000000000"
                         :account-id "account_id"
                         :type "basic"
                         :params {:title "My custom item"
                                  :image-url "www.example.com/image.png"
                                  :background-color "#FCF1EE"
                                  :body-color "#FCF1EE"
                                  :title-color "#333"
                                  :body "Some body text to display"}
                         :url "http://aan.io"}
```

### Webhooks
Web hooks allow your application to receive real-time, push notification of events in an account. 

You can register a webhook:

```clojure
(mondo/register-webhook access-token account-id webhook-url)
```

List the web hooks registered on an account:

```clojure
(mondo/list-webhooks access-token account-id)
```

Or delete a webhook to stop receiving transaction events:

```clojure
(mondo/delete-webhook access-token webhook-id)
```

### Attachments

The first step when uploading an attachment is to obtain a temporary URL to which the file can be uploaded. The response will include a `:file-url` which will be the URL of the resulting file, and an `:upload-url` to which the file should be uploaded to.
```clojure
(mondo/upload-attachment access-token file-name file-type)
```
Returns:
```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :file-url "https://s3-eu-west-1.amazonaws.com/mondo-image-uploads/user_00009237hliZellUicKuG1/LcCu4ogv1xW28OCcvOTL-foo.png"
 :upload-url "https://mondo-image-uploads.s3.amazonaws.com/user_00009237hliZellUicKuG1/LcCu4ogv1xW28OCcvOTL-foo.png?AWSAccessKeyId=AKIAIR3IFH6UCTCXB5PQ\u0026Expires=1447353431\u0026Signature=k2QeDCCQQHaZeynzYKckejqXRGU%!D(MISSING)"}
```

Once you have obtained a URL for an attachment, either by uploading to the `:upload-url` obtained from the `upload-attachment` endpoint above or by hosting a remote image, this URL can then be registered against a transaction. Once an attachment is registered against a transaction this will be displayed on the detail page of a transaction within the Mondo app.
 ```clojure
(mondo/register-attachment access-token external-id file-url file-type)
```
Returns:
```clojure
{:status {:success? true 
          :error-code 200 
          :error-name "OK" 
          :error-body "All is well."}
 :attachment {:id "attach_00009238aOAIvVqfb9LrZh"
              :user_id "user_00009238aMBIIrS5Rdncq9"
              :external-id "tx_00008zIcpb1TB4yeIFXMzx"
              :file-url "https://s3-eu-west-1.amazonaws.com/mondo-image-uploads/user_00009237hliZellUicKuG1/LcCu4ogv1xW28OCcvOTL-foo.png"
              :file-type "image/png"
              :created #inst "2015-12-01T00:00:00.000-00:00"}}

```

To remove an attachment, simply deregister this using its `id`
 ```clojure
(mondo/deregister-attachment access-token id)
```


## Questions & Shortcomings?
I don't have my Mondo account/card/app yet so this is pre-pre-alpha. That being said...

1. Would be really good to have an API version made available. Adding the request header `X-Api-Version` to make sure of  compatibility going forward.
2. It's unclear at this time whether Pagination allows for `before` to be either an instant or a transaction-id.


> Just need to get my hands on one of these now....
![](https://raw.githubusercontent.com/adamneilson/mondo-clj/master/resources/mondo-alpha-card.png)

## License

Copyright © 2015 [Adam Neilson](http://aan.io)

Distributed under the [Eclipse Public License](http://www.eclipse.org/legal/epl-v10.html) either version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2015-11-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2015-11-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/mondo-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mondo-clj/compare/0.1.0...0.1.1


test/mondo_clj/core_test.clj
(ns mondo-clj.core-test
  (:require [clojure.test :refer :all]
            [mondo-clj.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/mondo_clj/api.clj
(ns mondo-clj.api
  (:require 
    [mondo-clj.util :refer :all]
    [org.httpkit.client :as http]
    [cheshire.core :refer :all]))


;(def api-url "https://api.getmondo.co.uk")
(def api-url "https://production-api.gmon.io")



(def basic-request-options {:user-agent "mondo-clj-lib"
                            :headers {"X-Api-Version" "unknown"}
                            :keepalive 3000          ; Keep the TCP connection for 3000ms
                            :timeout 1000      ; connection timeout and reading timeout 1000ms
                            :filter (http/max-body-filter (* 1024 1000)) ; reject if body is more than 1000k
                            :insecure? false ; Need to contact a server with an untrusted SSL cert?
                            :follow-redirects false})



(defn get-error [error-code]
  (cond
    (= error-code 200) {:success? true :error-code 200 :error-name "OK" :error-body "All is well."}
    (= error-code 400) {:success? false :error-code 400 :error-name "Bad Request" :error-body "Your request has missing parameters or is malformed."}
    (= error-code 401) {:success? false :error-code 401 :error-name "Unauthorized" :error-body "Your request is not authenticated."}
    (= error-code 403) {:success? false :error-code 403 :error-name "Forbidden" :error-body "Your request is authenticated but has insufficient permissions."}
    (= error-code 405) {:success? false :error-code 405 :error-name "Method Not Allowed" :error-body "You are using the incorrect HTTP verb. Double check whether it should be POST/GET/DELETE/etc."}
    (= error-code 404) {:success? false :error-code 404 :error-name "Page Not Found" :error-body "The endpoint requested does not exist."}
    (= error-code 406) {:success? false :error-code 406 :error-name "Not Acceptable" :error-body "Your application does not accept the content format returned according to the Accept headers sent in the request."}
    (= error-code 429) {:success? false :error-code 429 :error-name "Too Many Requests" :error-body "Your application is exceeding its rate limit. Back off, buddy. :p"}
    (= error-code 500) {:success? false :error-code 500 :error-name "Internal Server Error" :error-body "Something is wrong on our end. Whoopsie."}
    (= error-code 504) {:success? false :error-code 504 :error-name "Gateway Timeout" :error-body "Something has timed out on our end. Whoopsie."}
    :else (throw (Throwable. "An unknown error has occurred."))))
    


(defn GET 
  ([uri access-token] (GET uri access-token {}))
  ([uri access-token params]
   (let [url (str api-url uri)
         options (merge basic-request-options {:url url
                                               :method :get
                                               :oauth-token (when (seq access-token) access-token)
                                               :query-params params})
         {:keys [status body error]} @(http/get url options)]
     (if error
       (get-error status)
       (-> (decode body)
           (assoc :status (get-error status))
           (keywordize-map))))))




(defn POST 
  ([uri access-token] (POST uri access-token {}))
  ([uri access-token params]
   (let [url (str api-url uri)
         options (merge basic-request-options {:url url
                                               :method :post
                                               :oauth-token (when (seq access-token) access-token)
                                               :form-params params})
         {:keys [status body error]} @(http/post url options)]
     (if error
       (get-error status)
       (-> (decode body)
           (assoc :status (get-error status))
           (keywordize-map))))))


(defn PATCH [uri access-token metadata]
  (let [url (str api-url uri)
        ;"metadata[$key1]=$value1" 
        options (merge basic-request-options 
                       {:url url
                        :method :patch
                        :oauth-token (when (seq access-token) access-token)
                        :form-params metadata})
        {:keys [status body error]} @(http/post url options)]
    (if error
      (get-error status)
      (-> (decode body)
          (assoc :status (get-error status))
          (keywordize-map)))))




(defn DELETE [uri access-token]
  (let [url (str api-url uri)
        options (merge basic-request-options 
                       {:url url
                        :method :delete
                        :oauth-token (when (seq access-token) access-token)})
        {:keys [status body error]} @(http/delete url options)]
    (if error
      (get-error status)
      (-> (decode body)
          (assoc :status (get-error status))
          (keywordize-map)))))





src/mondo_clj/util.clj
(ns mondo-clj.util)



(defn keywordize-map [m] (reduce (fn [m [k v]] 
                                   (assoc m (-> k
                                                (str)
                                                (clojure.string/trim)
                                                (clojure.string/replace #"_" "-")
                                                (keyword)) v)) {} m))


(defn underscore-keys [m] (reduce (fn [m [k v]] 
                                   (assoc m (-> k
                                                (str)
                                                (clojure.string/trim)
                                                (clojure.string/replace #"-" "_")
                                                (keyword)) v)) {} m))


(defn remove-nils 
  "Removes all nil valued kv pairs from a map" 
  [m]
  (apply dissoc m (for [[k v] m :when (nil? v)] k)))




(defn prepare-map [m]
  (-> m
      (remove-nils)
      (underscore-keys)))




(def not-nil? (complement nil?))


(defn is-valid-txn-id? "Does the format of the txn id look valid?" [transaction-id]
  (not-nil? (re-find transaction-id #"tx\_[A-Za-z0-9]*")))


(defn hex-colour "doc-string" [s]
  (when (string? s)
    (re-find #"\#[a-fA-F0-9]{3,6}" s)))


(defn pre-process-vals 
  "Do any pre-processing or transformations on map." 
  [m]
  (reduce (fn [m [k v]] 
            (assoc m k (if (string? v) 
                         (clojure.string/trim v)
                         v))) {} m))

(defn is-between? 
  "Is the value between the high and low vals (inclusive)." 
  [value low high]
  (and (>= value low)
       (<= value high)))


(defn instant-to-zulu "doc-string" [inst]
  (if (= (type inst) java.util.Date)
    (.format (java.text.SimpleDateFormat. 
               "yyyy-MM-dd'T'HH:mm:ss'Z'") inst)
    inst))

(defn zulu-to-instant "doc-string" [zulu]
  (.parse (java.text.SimpleDateFormat. "yyyy-MM-dd'T'HH:mm:ss") 
          (-> zulu (clojure.string/split #"\.") first)))


(defn deep-merge
  "Recursively merges maps. If keys are not maps, the last value wins."
  [& vals]
  (if (every? map? vals)
    (apply merge-with deep-merge vals)
    (last vals)))


(defn round2
  "Round a double to the given precision (number of significant digits)"
  [d precision]
  (let [factor (Math/pow 10 precision)]
    (/ (Math/round (* d factor)) factor)))


(defn coerce-to-double-monetary-amount 
  "doc-string" 
  [i]
  (-> i
      (* 0.01)
      (round2 2)))


src/mondo_clj/core.clj
(ns mondo-clj.core
  (:require 
    [mondo-clj.util :refer :all]    
    [mondo-clj.api :as api]))





;;================================================
;;
;;  AUTHORIZATION
;;
;;================================================


(defn get-access-token 
  "An access token is tied to both your application (the client) and 
  an individual Mondo user and is valid for several hours.

  Returns a map:

  {:access-token \"access_token\",
  :client-id \"client_id\",
  :expires-in 21600,
  :refresh-token \"refresh_token\",
  :token-type \"Bearer\",
  :user-id \"user_id\"}

  NB: Your client may only have one active access token at a time, per 
  user. Acquiring a new access token will invalidate any other token 
  you own for that user." 
  [{:keys [grant-type
           client-id
           client-secret
           username
           password]
    :or {grant-type "password"}}]
  ;check that the required fields are populated
  (if (every? not-nil? [grant-type client-id client-secret username password])
    (->> {:grant-type grant-type
          :client-id client-id
          :client-secret client-secret
          :username username
          :password password}
         (pre-process-vals)
         (prepare-map)
         (api/POST "/oauth2/token"))
    (api/get-error 400)))



(defn refresh-access-token 
  "To limit the window of opportunity for attackers in the 
  event an access token is compromised, access tokens expire 
  after 6 hours. To gain long-lived access to a user’s account,
  it’s necessary to \“refresh\” your access when it expires using 
  a refresh token. Only \“confidential\” clients are issued refresh 
  tokens – \“public\” clients must ask the user to re-authenticate.

  Returns:

  {
  :access_token \"access_token_2\"
  :client_id \"client_id\"
  :expires_in 21600
  :refresh_token \"refresh_token_2\"
  :token_type \"Bearer\"
  :user_id \"user_id\"
  }

  Refreshing an access token will invalidate the previous token, if 
  it is still valid. Refreshing is a one-time operation." 
  [{:keys [grant-type
           client-id
           client-secret
           refresh-token]
    :or {grant-type "refresh_token"}}]
  (if (every? not-nil? [grant-type client-id client-secret refresh-token])
    (->> {:grant-type grant-type
          :client-id client-id
          :client-secret client-secret
          :refresh-token refresh-token}
         (pre-process-vals)
         (prepare-map)
         (api/POST "/oauth2/token"))
    (api/get-error 400)))



(defn whoami 
  "To get information about an access token, you can call the 
  /ping/whoami endpoint." 
  [access-token]
  (if (not-nil? access-token)
    (api/GET "/ping/whoami" access-token)
    (api/get-error 400)))



;;================================================
;;
;;  ACCOUNTS
;;
;;================================================


(defn list-accounts 
  "Returns a list of accounts owned by the currently authorised 
  user." 
  [access-token]
  (if (not-nil? access-token)
    (let [result (api/GET "/accounts" access-token)]
      (assoc-in result [:accounts] 
                (map (fn [acc] 
                       (update-in acc [:created] zulu-to-instant))
                     (:accounts result))))
    (api/get-error 400)))




;;================================================
;;
;;  BALANCE
;;
;;================================================

(defn read-balance
  "Returns balance information for a specific account."
  [access-token account-id]
  (if (every? not-nil? [access-token account-id])
    (let [result (->> {:account-id account-id}
                      (pre-process-vals)
                      (prepare-map)
                      (api/GET "/balance" access-token))]
      (if (and (contains? result :balance)
               (contains? result :spend-today))
        (-> result
            ; coerce the balance & spend-today to doubles
            (update-in [:balance] coerce-to-double-monetary-amount)
            (update-in [:spend-today] coerce-to-double-monetary-amount))
        ; we didn't get the balance in the result
        result))
    (api/get-error 400)))

       




;;================================================
;;
;;  TRANSACTIONS
;;
;;================================================



(defn get-transaction 
  "Get a specific txn" 
  [access-token transaction-id]
  (if (and (every? not-nil? [access-token transaction-id])
           (is-valid-txn-id? transaction-id))
    (-> (api/GET (format "/transactions/%s" transaction-id) access-token {"expand[]" "merchant"})
        (update-in [:transaction :created] zulu-to-instant)
        (update-in [:transaction :merchant :created] zulu-to-instant)
        (update-in [:transaction :account-balance] coerce-to-double-monetary-amount)
        (update-in [:transaction :amount] coerce-to-double-monetary-amount)
        )
    (api/get-error 400)))



(defn list-transactions 
  "List transactions that were made against a specific account. " 
  [{:keys [access-token
           account-id
           limit
           since
           before]
    :or {limit 100
         before (java.util.Date.)}}]
  (if (and (every? not-nil? [access-token account-id])
           (is-between? limit 0 100))
    (let [opts (promise)
          before-zulu (instant-to-zulu before)
          since-zulu (instant-to-zulu since)]

      ; 'since' is a weird one as it can be either an 
      ; instant OR a txn ID. Without the API to hit this
      ; is a bit of guess work at the moment. Pagination
      ; does feel a little clunky...

      ; is it set?
      (if (nil? since)
        ; moot point
        (deliver opts (prepare-map {:account-id account-id
                                    :limit limit 
                                    :before before-zulu}))
        ; test for txn ID
        (if (and (string? since)
                 (is-valid-txn-id? since))
          (deliver opts (prepare-map {:account-id account-id
                                      :limit limit 
                                      :since since}))

          ;however if it's an instant then we may want to have 
          ;a timeframe with the before instant as well...
          (deliver opts (prepare-map {:account-id account-id
                                      :limit limit 
                                      :since since-zulu 
                                      :before before-zulu}))))

      ; covert all zulu times to instants
      (let [result (api/GET "/transactions" access-token @opts)]
        (-> result
            (assoc-in [:transactions] (map (fn [tx] (update-in tx [:created] zulu-to-instant)) (:transactions result)))
            (assoc-in [:transactions] (map (fn [tx] (update-in tx [:account-balance] coerce-to-double-monetary-amount)) (:transactions result)))
            (assoc-in [:transactions] (map (fn [tx] (update-in tx [:amount] coerce-to-double-monetary-amount)) (:transactions result)))
            )))
    (api/get-error 400)))


(comment 
  (prepare-map {:access-token "0000000000000000"
                       :account-id "account_id"
                       :limit [100 0] ;[per-page page-number]
                       :since #inst "2015-11-26T00:00:00.000-00:00"})
)





(defn annotate-transaction 
  "Set a metadata-map key's value as empty to delete it" 
  [access-token transaction-id metadata-map]
  (if (and (every? not-nil? [access-token transaction-id])
           (is-valid-txn-id? transaction-id)
           (> (count metadata-map) 0))
    (-> (api/PATCH (format "/transactions/%s" transaction-id) 
                   access-token 
                   (reduce 
                     (fn [m [k v]] 
                       (assoc m (format "metadata[%s]" (name k)) v))
                     {} metadata-map))
        (update-in [:transaction :created] zulu-to-instant)
        (update-in [:transaction :account-balance] coerce-to-double-monetary-amount)
        (update-in [:transaction :amount] coerce-to-double-monetary-amount))
    (api/get-error 400)))





;;================================================
;;
;;  FEED
;;
;;================================================



(defn create-feed-item
  "Inject an item/event into an accounts feed"
  [{:keys [access-token
           account-id
           type
           params
           url]
    :or {type "basic"}}]
  (if (and (every? not-nil? [access-token account-id type params])
           (every? not-nil? (vals (select-keys params [:title :image-url]))))
    (let [parameters (-> (select-keys params [:title
                                              :image-url
                                              :background-color
                                              :body-color
                                              :title-color
                                              :body])
                         (assoc :background-color (hex-colour (:background-color params)))
                         (assoc :body-color (hex-colour (:body-color params)))
                         (assoc :title-color (hex-colour (:title-color params)))
                         (remove-nils)
                         (prepare-map))]
      (api/POST "/feed" 
                access-token 
                (prepare-map {:account-id account-id
                              :type type
                              :params parameters
                              :url url})))
    (api/get-error 400)))

(comment (prepare-map {:access-token "0000000000000000"
                       :account-id "account_id"
                       :type "basic"
                       :params {:title "My custom item"
                                :image-url "www.example.com/image.png"
                                :background-color "#FCF1EE"
                                :body-color "#FCF1EE"
                                :title-color "#333"
                                :body "Some body text to display"}
                       :url "http://aan.io"}))




;;================================================
;;
;;  WEBHOOKS
;;
;;================================================

(defn register-webhook [access-token account-id url] 
  (if (every? not-nil? [access-token account-id url])
    (api/POST "/webhooks" 
              access-token 
              (prepare-map {:account-id account-id
                            :url url})))
  (api/get-error 400))


(defn list-webhooks [access-token account-id] 
  (if (every? not-nil? [access-token account-id url])
    (api/GET "/webhooks" 
             access-token 
             (prepare-map {:account-id account-id})))
  (api/get-error 400))



(defn delete-webhook [access-token webhook-id] 
  (if (every? not-nil? [access-token webhook-id])
    (api/DELETE (format "/webhooks/%s" webhook-id) access-token))
  (api/get-error 400))






;;================================================
;;
;;  ATTACHMENTS
;;
;;================================================

(defn upload-attachment [access-token file-name file-type]
  (if (every? not-nil? [access-token file-name file-type])
    (api/POST "/attachment/upload" 
              access-token 
              (prepare-map {:file-name file-name
                            :file-type file-type})))
  (api/get-error 400))



(defn register-attachment [access-token external-id file-url file-type]
  (if (and (every? not-nil? [access-token external-id file-name file-type])
           (is-valid-txn-id? external-id))
    (-> (api/POST "/attachment/register" 
                  access-token 
                  (prepare-map {:external-id external-id
                                :file-url file-url
                                :file-type file-type}))
        (update-in [:attachment :created] zulu-to-instant))
    (api/get-error 400)))

    

(defn deregister-attachment [access-token id]
  (if (every? not-nil? [access-token id])
    (api/POST "/attachment/deregister" 
              access-token 
              (prepare-map {:id id}))
    (api/get-error 400)))






